src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -1463,19 +1435,58 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
      * @param predicate the predicate to evaluate.
      * @return true if succesful.
      */
-    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Predicate<ItemStack> predicate)
+
+    private boolean tryTransferFromPosToWorkerIfNeeded(final BlockPos pos, @NotNull final Tuple<Predicate<ItemStack>, Integer> predicate)
     {
         final TileEntity entity = world.getTileEntity(pos);
         if (entity == null)
         {
             return false;
         }
 
-        return InventoryUtils.transferXOfFirstSlotInProviderWithIntoNextFreeSlotInItemHandler(
-          entity,
-          predicate,
-          Constants.STACKSIZE,
-          worker.getInventoryCitizen());
+        final int existingAmount = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(), predicate.getA());
+        int amount = 0;
+        if (predicate.getB() > existingAmount)
+        {
+            amount = predicate.getB() - existingAmount;
+        }
+        else
+        {
+            return true; // has already needed transfers...
+        }
+
+        int transferamount = 0;
+
+        do
```

Comment:
Please adapt the format. While on top.

Reply:
the do {} while is intended because it runs at least once.
a while {} do not.

Reply:
I change it because at this point it is working both ways